### ISSUE TITLE
Does isFilePasswordProtectedAtPath: leak resources?

### ISSUE THREAD

**Issue by hacknicity** [ZipArchive/ZipArchive#446]
Whilst investigating #445 I'm stepping through unfamiliar code and I'm suspicious that `isFilePasswordProtectAtPath:` could leak resources when it finds an encrypted file. I don't understand the unzip library enough to be sure, but just from inspecting the code there seems to be a mismatch in `unzOpenCurrentFile` and `unzCloseCurrentFile` calls 

~~~~
+ (BOOL)isFilePasswordProtectedAtPath:(NSString *)path {
    // Begin opening
    zipFile zip = unzOpen(path.fileSystemRepresentation);
    if (zip == NULL) {
        return NO;
    }
    
    int ret = unzGoToFirstFile(zip);
    if (ret == UNZ_OK) {
        do {
            ret = unzOpenCurrentFile(zip);
            if (ret != UNZ_OK) {
                return NO;
            }
            unz_file_info fileInfo = {};
            ret = unzGetCurrentFileInfo(zip, &fileInfo, NULL, 0, NULL, 0, NULL, 0);
            if (ret != UNZ_OK) {
                return NO;          // <<<< zip is not closed if this (perhaps very unlikely) error occurs
            } else if ((fileInfo.flag & 1) == 1) {
                return YES;       /// <<<< zip is not closed if an encrypted file is found
            }
            
            unzCloseCurrentFile(zip);
            ret = unzGoToNextFile(zip);
        } while (ret == UNZ_OK);
    }
    
    return NO;
}
~~~~

I suspect this will leak whatever internal resources that `unzOpenCurrentFile` creates when an archive is tested that is password-protected.

**Comment by hacknicity** [comment]
From inspection, it would seem that

~~~~
unzCloseCurrentFile(zip);
~~~~

could just be moved up above the `if` block so that `zip` was always closed before either failing because `unzGetCurrentFileInfo()` returned an error, or before returning `YES` because an encrypted file was found.

**Comment by hacknicity** [comment]
`isFilePasswordProtectedAtPath:` is in `SSZipArchive.m`.

**Comment by jhudsonWA** [comment]
Sorry, you are correct of course. I saw unzCloseCurrentFile and assumed. I'll take a look. Thanks!

**Comment by Coeur** [comment]
Fixed in 1584da8c39006555ba5e7e7ea6e3c5e9ed39c2d7
